packages/react-relay/relay-hooks/react-cache/useRefetchableFragmentInternal_REACT_CACHE.js

```diff
@@ -448,8 +455,8 @@ function useRefetchFunction<TQuery: OperationType>(
       // If the query needs an identifier value ('id' or similar) and one
       // was not explicitly provided, read it from the fragment data.
       if (
-        identifierField != null &&
-        !providedRefetchVariables.hasOwnProperty('id')
+        identifierQueryVariableName != null &&
```

Comment:
So, not quite. The warning below was already there before this PR. It was necessary because it doesn't only check for the existence of `identifierField`, rather, it makes sure we have an `identifierValue`, which is defined as `fragmentData[identifierField]`. 

Even if the compiler invariant holds and we _have_ a `identifierField` whenever there is an `identifierQueryVariableName`, `identifierValue` might be missing for a number of different reasons outside the control of the compiler. For example, there might be a malfunction in the Relay runtime or perhaps even in the users's GraphQL server.

Nevertheless, if we hit the warning we're definitely in a corrupted state and there's a case to be made for triggering an exception via an invariant, but I would you to confirm that this is indeed what we want since I'm not familiar with the team's heuristics for when to use a warning and when to go for an invariant.

Changing the compiler output as you're describing would still be nice. I'll see if I can get around to doing it.